keps/sig-apps/0010-20180522-deployment-inplace-update.md

```diff
@@ -0,0 +1,157 @@
+
+
+## Goals
+Enable deployment controller to update the containers inside the pods (such as container image, cmd etc.) without tearing down the pods
+
+## Proposal
+
+### API Change
+Augment deployment controller to have a new update strategy type: InPlaceUpdate.
+
+```
+
+type DeploymentStrategy struct {
+	// Type of deployment. Can be "Recreate" or "RollingUpdate" or "InPlaceUpdate". Default is RollingUpdate.
+	// +optional
+	Type DeploymentStrategyType
+
+	// Rolling update config params. Present only if DeploymentStrategyType =
+	// RollingUpdate.
+	//---
+	// TODO: Update this to follow our convention for oneOf, whatever we decide it
+	// to be.
+	// +optional
+	RollingUpdate *RollingUpdateDeployment
+
++	// InPlace update config params. Present only if DeploymentStrategyType =
++	// InPlaceUpdate.
++	// +optional
++	InPlaceUpdate *InPlaceUpdateDeployment
+}
+
+const(
+	// Kill all existing pods before creating new ones.
+	RecreateDeploymentStrategyType DeploymentStrategyType = "Recreate"
+
+        // Replace the old RCs by new one using rolling update i.e gradually scale down the old RCs and scale up the new one.
+        RollingUpdateDeploymentStrategyType DeploymentStrategyType = "RollingUpdate"
+    
+        // Update containers inside the pods without tearing down the pods
++	InPlaceUpdateDeploymentStrategyType DeploymentStrategyType = "InPlaceUpdate"
+)
+
+// Spec to control the desired behavior of in-place update
+type InPlaceUpdateDeployment struct {
+	// The number of pods to be updated in each batch.
+	// Value can be an absolute number (ex: 5) or a percentage of total pods at the start of update (ex: 10%).
+	// Absolute number is calculated from percentage by rounding down.
+	BatchSize intstr.IntOrString
+}
+```
+
+User specifies the new update strategy type in the spec and issue the request
+
+### Deployment Controller
+
+#### Update
+Below are two alternative approaches for in-place update in simplified words. 
+Wait until expected number of pods created, when deployment controller receives the in-place update request
+
+##### Approach 1
+
+```
+    newReplicaSet created with 0 replica
+    ...
+
+    // Find all old pods and update the unhealthy(failed, pending, not ready) pods first with the new spec from new ReplicaSet and then the running pods,
+    podsToUpdate = getPodsToUpdate(oldPods, BatchSize)
+    for pod range(podsToUpdate)
+        update(pod, newSpec)
+    
+    // Pause the old replica set. This is required for this scenario: if we reduce oldReplicaSet immediately, there's a chance
+    // that oldReplicaSet may kill the pods prematurely.
+    pause(oldReplicaSet)
+    
+    // Reduce oldReplicaSet by the BatchSize
+    update(oldReplicaSet, replicas - BatchSize)
+    
+    // update pod controllerRef to the new ReplicaSet
+    for pod range(podsToUpdate)
+        update(pod, newReplicaSet-controllerRef)
+
+    // Increase newReplicaSet by BatchSize
```

Comment:
Unfortunately this will not work.

1 - replica sets do not support pausing
2 - pausing (in the context of a deployment) means pausing an update. scale up/down still works as expected so if a new pod is added out of bound, it will get deleted automatically because now the deployment has more pods than expected.

Reply:
Correct, if we go with approach 1, then we need to make replica set support it. So far, I feel approach 2 is better to go, as it's simpler